domain/secret/state/state.go

```diff
@@ -306,6 +306,45 @@ ON CONFLICT(revision_uuid, name) DO UPDATE SET
 	return nil
 }
 
+// ListSecrets returns the secrets matching the specified criteria.
+// If all terms are empty, then all secrets are returned.
+func (st State) ListSecrets(ctx context.Context, uri *coresecrets.URI,
+	// TODO(secrets) - use all filter terms
+	revisions domainsecret.Revisions,
+	labels domainsecret.Labels, appOwners domainsecret.ApplicationOwners,
+	unitOwners domainsecret.UnitOwners, wantUser bool,
+) ([]*coresecrets.SecretMetadata, [][]*coresecrets.SecretRevisionMetadata, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+
+	var (
+		secrets        []*coresecrets.SecretMetadata
+		revisionResult [][]*coresecrets.SecretRevisionMetadata
+	)
+	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		var err error
+		secrets, err = st.listSecrets(ctx, tx, uri)
+		if err != nil {
+			return errors.Annotate(err, "querying secrets")
+		}
+		revisionResult = make([][]*coresecrets.SecretRevisionMetadata, len(secrets))
+		for i, secret := range secrets {
+			secretRevisions, err := st.listSecretRevisions(ctx, tx, secret.URI, nil)
```

Comment:
When you address the TODO above, please un-potato this so that we don't hit revisions per secret.